forta-uniswap-event-py/src/agent.py

```diff
@@ -0,0 +1,216 @@
+import binascii
+import json
+
+from forta_agent import Finding, FindingType, FindingSeverity
+from web3 import Web3
+import web3
+
+# Uniswap v2: Router 2
+# 0x7a250d5630b4cf539739df2c5dacb4c659f2488d
+ROUTER_ADDR = Web3.toChecksumAddress("0x7a250d5630b4cf539739df2c5dacb4c659f2488d")
+
+ETHER_THRESHOLD = Web3.toWei("5", "ether")
+
+CONTRACT_INST = None
+
+
+class AttrDict(dict):
+    """
+    Allows a dictionary to act as a class object and attributes can be retrieved with
+    the . "dot" notation or the "['']" square bracket notation. This is a very helpful helper
+    class since web3 typically uses square bracket notation and forta-agent uses dot syntax
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.__dict__ = self
+
+
+def get_contract_abi():
+    """
+    Given the address of a smart contract, return the abi provided by etherscan as a string
+    """
+    # API was retrieved from:
+    #   - https://etherscan.io/address/0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2#code
+    with open("weth_abi.json", "r") as f:
+        abi = json.loads(f.read())
+
+    return abi
+
+
+def get_contract_instance():
+    """
+    Get an instance of the contract by using the ABI saved in 'router_abi.json`
+    This will allow us to encode and decode function parameters
+    """
+    global CONTRACT_INST
+    if CONTRACT_INST:
+        return CONTRACT_INST
+
+    abi = get_contract_abi()
+    w3 = Web3()
+    CONTRACT_INST = w3.eth.contract(abi=abi)
+    return CONTRACT_INST
+
+
+def create_alert(to_addr, from_addr, amount_wad):
+    """
+    Return an alert with a metadata field that contains
+        - to address
+        - from address
+        - amount in wad
+    """
+    return Finding(
+        {
+            "name": "Uniswap swap detector",
+            "description": "Large swap on Uniswap detected",
+            "alert_id": "AE-UNISWAP-LARGESWAP-EVENT",
+            "type": FindingType.Suspicious,
+            "severity": FindingSeverity.Low,
+            "metadata": {"from": from_addr, "to": to_addr, "amount": amount_wad},
+            "everestId": "0xa2e07f422b5d7cbbfca764e53b251484ecf945fa",
+        }
+    )
+
+
+def handle_transaction(transaction_event):
+    """
+    Entry point for a transaction
+    """
+    input_data = transaction_event.transaction.data
+    if not input_data:
+        return []
+
+    # Ensure the 'to' field exists (will be a None on contract creation) and that it matches
+    # the ROUTER_ADDR
+    if (
+        transaction_event.transaction.to
+        and Web3.toChecksumAddress(transaction_event.transaction.to) != ROUTER_ADDR.lower()
```

Comment:
This looks like it will likely always be True.
Pick whether you want to stick with all checksum addresses or all lowercase addresses, but don't compare one to the other.

Reply:
Excellent catch!